src/components/application_manager/src/app_service_manager.cc

```diff
@@ -528,4 +532,256 @@ void AppServiceManager::AppServiceUpdated(
+        return true;
+      }
+
+    } break;
+
+    case MessageType::kResponse: {
+      LOG4CXX_DEBUG(logger_, "Handle response");
+      if (!rpc_message[strings::msg_params].keyExists(strings::result_code)) {
+        break;
+      }
+      auto result_code = static_cast<mobile_apis::Result::eType>(
+          rpc_message[strings::msg_params][strings::result_code].asInt());
+      if (result_code == mobile_apis::Result::UNSUPPORTED_REQUEST) {
+        LOG4CXX_DEBUG(logger_, "Service sent unsupported request");
+        CycleThroughRPCPassingRequests(correlation_id);
+        if (request_queue_[correlation_id].empty()) {
+          LOG4CXX_DEBUG(
+              logger_, "No services left in map. using core to handle request");
+          current_request_[correlation_id]
+              .message[strings::params][strings::connection_key] =
+              current_request_[correlation_id].origin_connection_key;
+          smart_objects::SmartObjectSPtr result =
+              std::make_shared<smart_objects::SmartObject>(
+                  current_request_[correlation_id].message);
+          request_queue_.erase(correlation_id);
+          current_request_.erase(correlation_id);
+          messages_using_core_.insert(correlation_id);
+          RemoveTimeoutTimer(correlation_id);
+          app_manager_.GetRPCService().ManageMobileCommand(
+              result, commands::Command::SOURCE_MOBILE);
+        } else {
+          LOG4CXX_DEBUG(logger_,
+                        "Forwarding request to service with app id "
+                            << current_request_[correlation_id]
+                                   .destination_connection_key);
+          smart_objects::SmartObjectSPtr result =
+              std::make_shared<smart_objects::SmartObject>(
+                  current_request_[correlation_id].message);
+          AddTimeoutTimer(correlation_id);
+          app_manager_.GetRPCService().SendMessageToMobile(result);
+        }
+        return true;
+      }
+
+      LOG4CXX_DEBUG(logger_, "Valid RPC passing response");
+
+      if (current_request_.find(correlation_id) != current_request_.end()) {
+        LOG4CXX_DEBUG(
+            logger_,
+            "Forwarding response to mobile app "
+                << current_request_[correlation_id].origin_connection_key);
+        rpc_message[strings::params][strings::connection_key] =
+            current_request_[correlation_id].origin_connection_key;
+        smart_objects::SmartObjectSPtr result =
+            std::make_shared<smart_objects::SmartObject>(rpc_message);
+
+        request_queue_.erase(correlation_id);
+        current_request_.erase(correlation_id);
+        RemoveTimeoutTimer(correlation_id);
+        app_manager_.GetRPCService().SendMessageToMobile(result);
+        return true;
+      }
+
+      return false;
+    } break;
+  }
+
+  return false;
+}
+
+bool RPCPassingHandler::CycleThroughRPCPassingRequests(
```

Comment:
`NextRequest`?

Reply:
I've renamed it to `CycleThroughRequests`. I think that's clearer than `NextRequest` since it just pops of the first element from the request queue

Reply:
`Cycle` sort of implies there is a loop involved, I think that something like `PerformNextRequest` would describe the functionality better, since you are just moving on to the next request in the list (The addition of `Perform` is because I had another comment about how this should send the next request in the queue as well, since that always seem to follow this function).

Reply:
Renaming it to `PerformNextRequest` and adding the code to send next request to core/mobile